specs/node_agent_flows_and_namespaces.adoc

```diff
@@ -0,0 +1,623 @@
+These stacks would be handled the same way as support for an storage system.
+
+
+== Use Cases
+
+This specification addresses the core node agent scope and impacts multiple
+specifications pertaining to the node agent capabilities:
+
+* Allow node agent to run on any node, including nodes where Tendrl components
+  are deployed. This would be achieved by tagging and tag based job routing,
+  https://github.com/Tendrl/specifications/issues/44[#44].
+* Automatic detection of services running on a node, including Tendrl and
+  storage system components,
+  https://github.com/Tendrl/specifications/issues/46[#46]. This functionality
+  also impacts the import cluster workflow to enable automatic detection of the
+  storage cluster version and layout,
+  https://github.com/Tendrl/specifications/issues/54[#54].
+* Provisioning support for ceph,
+  https://github.com/Tendrl/specifications/issues/48[#48], and gluster
+  https://github.com/Tendrl/specifications/issues/49[#49].
+* Various provisioning flows,
+  https://github.com/Tendrl/specifications/issues/47[#47],
+  https://github.com/Tendrl/specifications/issues/51[#51],
+  https://github.com/Tendrl/specifications/issues/52[#52].
+
+
+== Proposed Changes
+
+=== Platform Detection Modules
+
+Node agent must be able to indentify the base system type (centos, rhel, ubuntu
+etc.) and it's version. This information would later be used to determine which
+specific storage systems and/or their specific versions are supported and can
+thus be deployed on the platform.
+
+* Node agent needs to define a directory from which it can load the modules.
+* Each module should contain an interface method to be called to invoke it's
+  functionality.
+* Each module must contain the logic to ascertain only one specific platform
+  type and version.
+* Once a specific module detects a platform, the operation can be terminated
+  without invoking the remaining detection modules.
+* The supported platform types and versions would need to be hardcoded and
+  shipped with Tendrl.
+
+=== Storage System Integration Modules
+
+It is necessary to ship a part of the officially supported storage system
+specific integration namespace with the node agent itself. However, this should
+only be a "bootstrap" namespace, necessary to detect and provision the rest of
+the namespace and components.
+
+* By default, two types of integration modules need to be shipped with the node
+  agent:
+** Service detection modules, which have been covered in the <<Service Detection Modules>> section.
+** Provisioning modules to deploy the integration bridges or provisioning
+   systems (gdeploy, ceph-ansible etc.) for a specific storage system, which
+   have been covered in the <<Provisioning Modules>> section.
+* For both types of modules, node agent needs to define a separate directory
+  from where the specific type of integration provisioning modules can be loaded.
+* The modules can be python classes with a specific interface method to be
```

Comment:
I feel this spec should list wire-frame of these interfaces in implementation section.